src/liboslexec/llvm_util.cpp

```diff
@@ -1530,6 +1544,109 @@ LLVM_Util::write_bitcode_file (const char *filename, std::string *err)
 
 
 
+bool
+LLVM_Util::ptx_compile_group (llvm::Module* lib_module, const std::string& name,
+                              std::string& out)
+{
+    std::string new_triple = "nvptx64-nvidia-cuda";
+    std::string new_layout = "e-i64:64-v16:16-v32:32-n16:32:64";
+
+    // Clone the source Module to prevent the linker from tinkering with it
+    std::unique_ptr<llvm::Module> mod_ptr = llvm::CloneModule (module());
+
+    llvm::Module* linked_module = new_module (name.c_str());
+
+    // Make sure the module targets/layouts match
+    mod_ptr.get()->setTargetTriple (new_triple);
+    mod_ptr.get()->setDataLayout (new_layout);
+    linked_module->setTargetTriple (new_triple);
+    linked_module->setDataLayout (new_layout);
+
+    // First, link in the support library
+    std::unique_ptr<llvm::Module> lib_ptr (lib_module);
+    bool failed = llvm::Linker::linkModules (*linked_module, std::move (lib_ptr));
+    if (failed) {
+        ASSERT (0 && "Unable to link library module");
+    }
+
+    // Second, link in the compiled ShaderGroup module
+    failed = llvm::Linker::linkModules (*linked_module, std::move (mod_ptr));
+    if (failed) {
+        ASSERT (0 && "Unable to link group module");
+    }
+
+    // Verify that the NVPTX target has been initialized
+    std::string error;
+    const llvm::Target* llvm_target =
+        llvm::TargetRegistry::lookupTarget (new_triple, error);
+    if(! llvm_target) {
+        ASSERT (0 && "LLVM Target is not initialized");
+    }
+
+    llvm::TargetOptions  options;
+    options.AllowFPOpFusion                        = llvm::FPOpFusion::Standard;
+    options.UnsafeFPMath                           = 1;
+    options.NoInfsFPMath                           = 1;
+    options.NoNaNsFPMath                           = 1;
+    options.HonorSignDependentRoundingFPMathOption = 0;
+    options.FloatABIType                           = llvm::FloatABI::Default;
+    options.AllowFPOpFusion                        = llvm::FPOpFusion::Fast;
+    options.NoZerosInBSS                           = 0;
+    options.GuaranteedTailCallOpt                  = 0;
+    options.StackAlignmentOverride                 = 0;
+    options.UseInitArray                           = 0;
+
+    llvm::TargetMachine* target_machine = llvm_target->createTargetMachine(
+        new_triple, "sm_35", "+ptx50", options,
+        llvm::Reloc::Static, llvm::CodeModel::Default, llvm::CodeGenOpt::Aggressive);
```

Comment:
FYI, this line worked when built against LLVM 5.0, but against llvm 6.0 it says that there is no member named 'Default' in namespace 'llvm::CodeModel'.

Reply:
Ah, I see that the 'Default' enum value was removed in 6.0. I believe that ``CodeModel::Small`` should do the right thing in LLVM 4.0 through 7.0. I've tested it in 4, 5, and 6.